src/engine/subscription/Subscription.spec.ts

```diff
@@ -0,0 +1,575 @@
+          description: 'Just some description',
+        },
+      },
+      subscriptions: [
+        new Subscription({
+          name: 'SomeSubWithPreProcessor',
+          type: SUBSCRIPTION_TYPE_CREATE,
+          description: 'build item',
+          attributes: ['someAttribute'],
+          delimiter: '/',
+          // wildCard: '',
+          // pattern: '',
+          preProcessor: (
+            _entity,
+            _id,
+            _source,
+            input,
+            typeName,
+            entitySubscription,
+          ) => {
+            if (entitySubscription.attributes && Object.keys(input).length) {
+              const delimiter = entitySubscription.delimiter;
+              const filled = entitySubscription.attributes
+                .map(attribute => input[attribute])
+                .reduce((acc, curr) => `${acc + delimiter + curr}`, '');
+
+              const topic = `${entitySubscription.name}${_entity.name}${filled}`;
+              return topic;
+            }
+
+            return null;
+          },
+        }),
+      ],
+    });
+
+    const someAttribute = 'test';
+
+    it('should pass through preProcessor if it is declared', async () => {
+      const subscriptionByName = testEntity.getSubscriptionByName(
+        'SomeSubWithPreProcessor',
+      );
+      expect(subscriptionByName).toMatchSnapshot();
+
+      const setup = await generateTestSchema({ entities: [testEntity] });
+      const graphqlSchema = generateGraphQLSchema(setup.configuration);
+      const subscriptionDoc = parse(`
+        subscription someSubWithPreProcessorSomeTestsEntityName($input: SomeSubWithPreProcessorSomeTestsEntityNameInput!) {
+          someSubWithPreProcessorSomeTestsEntityName(input: $input) {
+            someTestsEntityName {
+              # someAttribute
+              anotherAttribute
+            }
+          }
+        }`);
+
+      // subscription topic will be automatically generated by shyft
+      let subscription = (await subscribe({
+        schema: graphqlSchema,
+        document: subscriptionDoc,
+        variableValues: {
+          input: {
+            someTestsEntityName: {},
+          },
+        },
+        contextValue: { pubsub },
+      })) as AsyncIterableIterator<any>;
+
+      let pending = subscription.next();
+
+      await pubsub.publish('SomeSubWithPreProcessorSomeTestsEntityName', {
+        // someTestsEntityName: {
+        someAttribute,
+        anotherAttribute: 'world',
+        // },
+      });
+
+      let result = await pending;
+      expect(result).toMatchSnapshot('withoutPreProcessorResult');
+      expect(await subscription.return()).toMatchSnapshot(
+        'withoutPreProcessorEnd',
+      );
+
+      // subscription topic will be generated in preProcessor
+      subscription = (await subscribe({
+        schema: graphqlSchema,
+        document: subscriptionDoc,
+        variableValues: {
+          input: {
+            someTestsEntityName: {
```

Comment:
By providing an attribute (someAttribute), we'll trigger the preProcessor condition and create custom topic.